### ISSUE TITLE
[🐛 Bug]: TypeScript types for driver.wait return single item instead of Array

### ISSUE THREAD

**Issue by slhck** [SeleniumHQ/selenium#14239]
### What happened?

I am using:

- @types/selenium-webdriver: 4.1.24
- selenium-webdriver: 4.22.0
- typescript: 5.5.3

The TypeScript definition for `webdriver.wait` seems to be wrong or inconsistent.

In the below code snippet, I get an error:

> `Property 'length' does not exist on type 'WebElement'.ts(2339)`

Yet, `result` is an array of `WebElement`s:

```
> result.at(0)
WebElement {driver_: Driver, id_: Promise, log_: Logger}
```

Since `driver.wait` returns a `WebElementPromise`, it assumes it's only a single item, whereas the `until.elementsLocated` returns a `Condition<WebElement[]>`.

### How can we reproduce the issue?

```typescript
import webdriver, { By, until } from "selenium-webdriver";
import chrome from "selenium-webdriver/chrome.js";

const chromeOptions = new chrome.Options().addArguments(
  "--mute-audio",
  "--autoplay-policy=no-user-gesture-required"
);

const driver = new webdriver.Builder()
  .forBrowser("chrome")
  .setChromeOptions(chromeOptions)
  .build();

await driver.get("https://www.youtube.com");

const primaryBtnsPromise = driver.wait(
  until.elementsLocated(By.css("tp-yt-paper-button.style-primary"))
);
const rejectButtonPromise = driver.wait(
  until.elementsLocated(By.css('button[aria-label*="Reject"]'))
);

const result = await Promise.race([primaryBtnsPromise, rejectButtonPromise]);

console.log("result", Array.isArray(result)); // --> true!

if (result.length <= 0) {
  throw new Error(`No buttons found!`);
}
```

### Relevant log output

```shell
See above
```

### Operating System

N/A

### Selenium version

4.22.0

### What are the browser(s) and version(s) where you see this issue?

N/A

### What are the browser driver(s) and version(s) where you see this issue?

N/A

### Are you using Selenium Grid?

No

**Comment by slhck** [comment]
If I knew which change caused that regression I could have a deeper look, but the types and inner workings are bit too complex for me to grasp as an outsider.

**Comment by marlapativ** [comment]
@slhck 
One workaround is to force the constraints to use WebElement[]

```javascript
const primaryBtnsPromise = driver.wait<WebElement[]>(
  until.elementsLocated(By.css("tp-yt-paper-button.style-primary"))
);
const rejectButtonPromise = driver.wait<WebElement[]>(
  until.elementsLocated(By.css('button[aria-label*="Reject"]'))
);
```

From what I understand, the issue shows up because typescript cannot differentiate between the following for `wait` function 
1. WebElementCondition (`Condition<WebElement>`)
2. `Condition<WebElement[]>`

**Comment by slhck** [comment]
Yes, that might work. The thing is, this seems to be a regression. I have only encountered this error after upgrading from an older version of the types/selenium.

**Comment by marlapativ** [comment]
@diemol, Can I pick this issue?

Also if I understand correctly, I'd need to do changes in the types [`DefinitelyTyped repo`](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/selenium-webdriver), Correct?

**Comment by harsha509** [comment]
Hi @marlapativ,

Yes, changing it there is correct.

I've looked into the types, and it seems new types have been merged recently. However, using generics is still the best approach for now. as returning `Promise<WebElement | WebElement[]>` from wait requires type guarding in code `Array.isArray(result)`

For WebElements, you should use `driver.wait<WebElement[]>` as it automatically infers the correct type for `WebElement`.

Thanks,
Sri

**Comment by slhck** [comment]
Is there any indication as to which changes introduced this regression? I am a bit lost in the number of files and commits.

But yes, the return value of `driver.wait` should be generically inferred from the thing it's waiting for, as `until.elementsLocated` returns an array, always.

**Comment by slhck** [comment]
Any updates on this?

**Comment by jonhermansen** [comment]
The bug exists in @types/selenium-webdriver version 4.1.23, published on 2024-05-31T17:06:57.036Z.

The bug does not exist in 4.1.22, published on 2024-03-07T23:35:52.529Z.

So @slhck, it makes sense that you observed a problem with compilation around the time you filed the bug. I tried looking at the diff between the two packages, but the problem is not immediately obvious at first glance.

**Comment by jonhermansen** [comment]
After git-bisecting DefinitelyTyped repo, I found the offending commit: https://github.com/DefinitelyTyped/DefinitelyTyped/commit/b0f70c1ea223dc166a34a5fab06f350a2f87c96e

**Comment by slhck** [comment]
It seems the PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/71956 didn't land, so I'm going to keep this open.

**Comment by slhck** [comment]
I'm trying again with my own PR here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/74149

**Comment by slhck** [comment]
This was merged in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/74149 – closing here.
